### PR TITLE
Adding exceptions classes

### PR DESCRIPTION
**Pull request by JoeZiminski**
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/brainglobe/.github/blob/main/CONTRIBUTING.md).

## Description

Begin to implement custom exceptions as mentioned in brainglobe/brainreg#82. Started by a custom exception for failed file loading. Currently checks whether the user is trying to pass a directory containing a single .tiff (e.g. 3D .tiff), in which case gives a message to pass the path including filename. Otherwise, it will print the origional traceback with an additional line indicating the problem is with file loading.

Some points to discsuss:
- it could be worth pre-checking inputs and raising an error there, rather than wait to fail on file loading, although this might come with performance penality.
- In the traceback, all attempted image files could be loaded e.g. with PIL (require new dependency), to check their size - this way a tailored traceback can be given.
- A couple of functions in this class could be moved to a more general utils module
- it would be useful to have a config of accepted file extensions (e.g. .tif, .tiff)

**What is this PR**

- [ ] Bug fix
- [X ] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
see issue brainglobe/brainreg#82

**What does this PR do?**
Implements custom exception class for load file errors

## References

Please reference any existing issues/PRs that relate to this PR.
brainglobe/brainreg#82

## How has this PR been tested?
pytest suite

Please explain how any new code has been tested, and how you have ensured that no existing functionality has changed.
pytest suite, and the new code is only called when file loading fails (i.e. when there is already an issue)

## Is this a breaking change?
No

If this PR breaks any existing functionality, please explain how and why.
NA

## Does this PR require an update to the documentation?
No

If any features have changed, or have been added. Please explain how the documentation has been updated (and link to the associated PR). See [here](https://docs.cellfinder.info/for-developers/documentation) for details.

## Checklist:

- [X ] The code has been tested locally
- [ ] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [X ] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by adamltyson** [comment]
@JoeZiminski is this ready to review?

File: tests/tests/test_unit/test_exceptions.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+import pytest
+
+from brainreg.cli import main as brainreg_run
+from brainreg.exceptions import LoadFileException
+
+test_data_dir = Path(os.getcwd()) / "tests" / "data"
+
+one_2d_tiff_data_dir = test_data_dir / "input" / "exceptions" / "one_2d_tiff"
+one_3d_tiff_data_dir = test_data_dir / "input" / "exceptions" / "one_2d_tiff"
```
Comment (adamltyson) [inline tests/tests/test_unit/test_exceptions.py:13]: Presumably this should refer to the 3D directory?

File: tests/tests/test_unit/test_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    assert (
+        f"File at {mismatched_dims_data_dir} "
+        f"failed to load. Ensure all image files "
+        f"contain the same number of pixels. Full traceback above."
+        in e.value.message
+    )
+
+
+def test_one_2d_tiff_error(test_output_dir):
```
Comment (adamltyson) [inline tests/tests/test_unit/test_exceptions.py:62]: This test passes for me, but when I run a similar example locally, the exception isn't raised, but a niftyreg error is later on.

**Review by adamltyson — CHANGES_REQUESTED** [review]

**Comment by adamltyson** [comment]
I've left a couple of comments. I also think that the `test_unit` dir should contain an `__init__.py` file. I seem to remember this was needed for reporting code coverage. We should be consistent anyway (`test_integration` has it).

**Comment by JoeZiminski** [comment]
Thanks pushed commit fixing those issues, will check out that failing test case, unsure why that is

**Comment by JoeZiminski** [comment]
> I've left a couple of comments. I also think that the `test_unit` dir should contain an `__init__.py` file. I seem to remember this was needed for reporting code coverage. We should be consistent anyway (`test_integration` has it).

In the dataset used in tests, when the single tiff is loaded, imio does not fail but the size of the output == 0 (which is checked). In the tutorial dataset, the image is loaded sucessfully as a (1, 154, 217) size array. Would it be sensible to a) check for cases that the first dimension == 1, and also raise error? Or, b) try - except niftyreg, and then in the exception class check if the first dimension == 1? I guess the first option is neater, but maybe there are cases when we want to load a single-dimension tiff?

**Comment by adamltyson** [comment]
I'd go with the first option. `brainreg` is a 3D registration library, I'm happy to not try to support 2D data in this way.

**Comment by JoeZiminski** [comment]
Added the single dimension case as ```np.min(target_brain.shape) == 1``` as I didn't want to make assumption singleton is on the first dimension, I guess we never want a case where a dimension has N = 1 anyways. But, if the volumne dimension is always in the first position it might be neater just to check that explicitly ```target_brain.shape[0] == 1```.  It turns out with the allen 100um, the files load with array size == 0, but for allen 50um they load with 3D array (1 singletom dim), I guess because of the scaling. 

For the new test case, used the allen 50um to make sure both conditions are tested. However, I suppose this adds another test dependency (downloading the 50 um atlas)

**Comment by adamltyson** [comment]
@JoeZiminski when I run these with example data, I don't think I get the correct errors:

```
 brainreg tests/data/input/exceptions/one_2d_tiff/loads_1_dim/image_0000.tif ~/Desktop -v 50 50 50 --orientation psl
```

gives:
```
File at tests/data/input/exceptions/one_2d_tiff/loads_1_dim/image_0000.tif failed to load. Ensure all image files contain the same number of pixels. Full traceback above.
```

and

```
brainreg tests/data/input/exceptions/one_2d_tiff/loads_no_dim/ ~/Desktop -v 50 50 50 --orientation psl
```

Produces a NiftyReg error

**Comment by JoeZiminski** [comment]
As discussed, just check whether a directory with a single file is attempted to be loaded. If so, raise error.

Otherwise, use imio to try and load. If this fails, inform the user this is where it failed and indicate they need to include
a folder of .tiff file where all pixels are the same size. 

Hopefully this will avoid the issues above and be much more robust. I tested this on the files I have but please let me know if anything doesn't look right, or the exception messages are not clear. 

Currently there are two functions in exceptions.py that could probably go somewhere else, I think we did discuss putting them elsewhere in the brainglobe ecosystem.

**Comment by adamltyson** [comment]
Thanks @JoeZiminski.

At some point I think these functions could go into [imio](https://github.com/brainglobe/imio), but for now, I'm happy to merge.

**Pull request merged**